pkg/prototype/flags_test.go

```diff
@@ -109,3 +112,194 @@ func TestBindFlags_duplicate_optional_param(t *testing.T) {
 	_, err := BindFlags(p)
 	require.Error(t, err)
 }
+
+func TestExtractParameters(t *testing.T) {
+	validPrototype := &Prototype{
+		APIVersion: "0.1",
+		Name:       "io.ksonnet.pkg.configMap",
+		Params: ParamSchemas{
+			{
+				Name:        "name",
+				Description: "Name to give the configMap",
+				Type:        String,
+			},
+			{
+				Name:        "data",
+				Description: "Data for the configMap",
+				Default:     strings.Ptr("{}"),
+				Type:        Object,
+			},
+			{
+				Name:        "val",
+				Description: "Value",
+				Default:     strings.Ptr("9"),
+				Type:        Number,
+			},
+		},
+	}
+
+	cases := []struct {
+		name      string
+		p         *Prototype
+		initFlags func(*testing.T, *Prototype, []string) *pflag.FlagSet
+		initFs    func(t *testing.T, fs afero.Fs)
+		args      []string
+		expected  map[string]string
+		isErr     bool
+	}{
+		{
+			name: "with valid flags",
+			p:    validPrototype,
+			initFlags: func(t *testing.T, p *Prototype, args []string) *pflag.FlagSet {
+				flags, err := BindFlags(p)
+				require.NoError(t, err)
+
+				err = flags.Parse(args)
+				require.NoError(t, err)
+
+				return flags
+			},
+			args: []string{
+				"--name=name",
+			},
+			expected: map[string]string{
+				"data": `{}`,
+				"name": `"name"`,
+				"val":  `9`,
+			},
+		},
+		{
+			name: "values from file",
+			p:    validPrototype,
+			initFlags: func(t *testing.T, p *Prototype, args []string) *pflag.FlagSet {
+				flags, err := BindFlags(p)
+				require.NoError(t, err)
+
+				err = flags.Parse(args)
+				require.NoError(t, err)
+
+				return flags
+			},
+			initFs: func(t *testing.T, fs afero.Fs) {
+				data := []byte(`{name: "name"}`)
+				afero.WriteFile(fs, "/values-file", data, 0644)
+			},
+			args: []string{
+				"--values-file=/values-file",
+			},
+			expected: map[string]string{
+				"data": `{}`,
+				"name": `"name"`,
+				"val":  `9`,
+			},
+		},
+		{
+			name: "missing a required flag",
+			p:    validPrototype,
+			initFlags: func(t *testing.T, p *Prototype, args []string) *pflag.FlagSet {
+				flags, err := BindFlags(p)
+				require.NoError(t, err)
+
+				err = flags.Parse(args)
+				require.NoError(t, err)
+
+				return flags
+			},
+			args:  []string{},
+			isErr: true,
+		},
+		{
+			name: "flag not defined",
+			p:    validPrototype,
+			initFlags: func(t *testing.T, p *Prototype, args []string) *pflag.FlagSet {
+				flags := pflag.NewFlagSet("prototype-flags", pflag.ContinueOnError)
+				return flags
+			},
+			args: []string{
+				"--name=name",
+				"--val=a",
```

Comment:
a name like `--undefined-flag` indicates where the error in this test will be better than `--val` here